core/transcoder.go

```diff
@@ -91,6 +150,28 @@ func (nv *NvidiaTranscoder) Stop() {
 	nv.session.StopTranscoder()
 }
 
+func runTranscodeLoop(gpu string, stack *segStack, workDir string) {
+	for {
+		seg := stack.pop()
```

Comment:
I don't think this loop will exit cleanly when the node stops while the stack is empty. One solution could be to send a cancellation signal via context. The top level context could be passed into `StartNvidiaTranscoders()` which passes it to `runTranscodeLoop()`.

Reply:
Yes, you're correct - the loop will never exit. This felt like a decent solution that got us 80% of the way with 20% of the effort, because there are only two times when the loop could be terminated: when a node shuts down gracefully, and when we run tests. Since we don't really attempt graceful termination elsewhere (and we have fallback / retry mechanisms for non-graceful termination), this felt OK.

There's a lot more incidental machinery if we bring in a cancellable context into `startTranscodeLoop`. In particular, it means "popping" from the stack actually means reading from a channel so we can `select` on the channel + context - so we still need to maintain the stack behavior internally, but write into a channel when a segment becomes available, rather than actually popping it. We'll also need some sort of feedback mechanism that the channel is available to be written to, since we don't want to queue/buffer segments since this is supposed to be a LIFO stack. All in all, the machinery adds up, and it feels a bit disproportionate to the cases it's meant to accommodate.

It's also more shuffling of data across channel boundaries - and we already send the data through a few channels between the load balancer, transcode loop, etc.

One alternative I've considered is to bring the cancellable context within the stack itself (and returning nil from the `pop` to break), but I haven't fully worked that out; probably by putting the queue in a `terminated` state once the context is signaled. If you still feel that we should cleanly shut down the loop regardless, then will look into it some more.

Reply:
Something like this, but it's still non-graceful - any segments that are in the stack when termination is signaled will be dropped. Still needs testing etc. 

```diff
diff --git a/cmd/livepeer/livepeer.go b/cmd/livepeer/livepeer.go
index b1f74219..229bd121 100644
--- a/cmd/livepeer/livepeer.go
+++ b/cmd/livepeer/livepeer.go
@@ -240,7 +240,7 @@ func main() {
 
        if *transcoder {
                if *nvidia != "" {
-                       core.StartNvidiaTranscoders(*nvidia, *datadir)
+                       core.StartNvidiaTranscoders(context.Background(), *nvidia, *datadir)
                        n.Transcoder = core.NewLoadBalancingTranscoder(*nvidia, core.NewNvidiaTranscoder)
                } else {
                        n.Transcoder = core.NewLocalTranscoder(*datadir)
diff --git a/core/transcoder.go b/core/transcoder.go
index 5bc891b4..64287f16 100644
--- a/core/transcoder.go
+++ b/core/transcoder.go
@@ -1,6 +1,7 @@
 package core
 
 import (
+       "context"
        "errors"
        "fmt"
        "io/ioutil"
@@ -78,16 +79,28 @@ type segStack struct {
        mu   *sync.Mutex
        cv   *sync.Cond
        gpu  string
+
+       terminated bool
 }
 
-func newSegStack(gpu string) *segStack {
+func newSegStack(ctx context.Context, gpu string) *segStack {
        mu := &sync.Mutex{}
-       return &segStack{
+       ss := &segStack{
                segs: []*nvSegData{},
                mu:   mu,
                cv:   sync.NewCond(mu),
                gpu:  gpu,
        }
+       go func() {
+               select {
+               case <-ctx.Done():
+                       mu.Lock()
+                       defer mu.Unlock()
+                       ss.terminated = true
+                       ss.cv.Broadcast()
+               }
+       }()
+       return ss
 }
 
 func (ss *segStack) push(seg *nvSegData) {
@@ -105,6 +118,9 @@ func (ss *segStack) pop() *nvSegData {
        defer ss.mu.Unlock()
        for len(ss.segs) <= 0 {
                ss.cv.Wait()
+               if ss.terminated {
+                       return nil
+               }
        }
        seg := ss.segs[len(ss.segs)-1]
        ss.segs = ss.segs[:len(ss.segs)-1]
@@ -129,11 +145,11 @@ func (nv *NvidiaTranscoder) Transcode(job string, fname string, profiles []ffmpe
 
 var nvidiaDevices map[string]*segStack
 
-func StartNvidiaTranscoders(devices string, workDir string) {
+func StartNvidiaTranscoders(ctx context.Context, devices string, workDir string) {
        nvidiaDevices = make(map[string]*segStack)
        d := strings.Split(devices, ",")
        for _, n := range d {
-               st := newSegStack(n)
+               st := newSegStack(ctx, n)
                nvidiaDevices[n] = st
                go runTranscodeLoop(st, workDir)
        }
@@ -153,6 +169,9 @@ func (nv *NvidiaTranscoder) Stop() {
 func runTranscodeLoop(stack *segStack, workDir string) {
        for {
                seg := stack.pop()
+               if nil == seg {
+                       return
+               }
                // Set up in / out config
                in := &ffmpeg.TranscodeOptionsIn{
                        Fname:  seg.fname,
```

Reply:
Ah I see what you mean now.

> All in all, the machinery adds up, and it feels a bit disproportionate to the cases it's meant to accommodate.

While I do think its worthwhile to get in the habit of keeping graceful termination in mind,  I agree that the effort/additional complexity involved with the extra machinery required in this case feels greater than the benefit of ensuring graceful termination. I'm ok with the code as is.

I did think a bit about a possible solution though. Maybe something like this:

```golang
type segStack struct {
    ...
    quit chan struct{}
}

// defer call this after calling StartNvidiaTranscoders()
func StopNvidiaTranscoders() {
    for _, device := range nvidiaDevices {
        close(device.quit)
        
        // If the stack is waiting, wake it up
        // Might need to do this in a for loop in case the the stack is not yet waiting
        // i.e. Code execution in the transcode loop is at stack.pop()
        device.sv.Signal()
    }
}

func runTranscodeLoop(stack *segStack, workDir string) {
    for {
        seg := stack.pop()

        select {
            case <-stack.quit:
                return
            default:
        }
    }
}
```

Feel free to look into the above if you'd like - I don't feel strongly one way or another.

Reply:
```golang
func runTranscodeLoop(stack *segStack, workDir string) {
    for {
        seg := stack.pop()

        select {
            case <-stack.quit:
                return
            default:
        }
```

As-is, isn't that still non-graceful? We could be exiting from the loop while the stack still has elements. Doable if we check the stack size and skip the select if nonzero, just requires some care in testing.

I'm hoping that this whole serialization workaround is temporary anyway, and we can revert as soon as LPMS perf is better.

Reply:
> As-is, isn't that still non-graceful? We could be exiting from the loop while the stack still has elements.

Ah didn't see your comment about dropping segments in the stack when termination is signaled when I posted my comment. I was thinking that dropping segments from the stack would be ok if termination is signaled by node shutdown.

> I'm hoping that this whole serialization workaround is temporary anyway, and we can revert as soon as LPMS perf is better.

SGTM 👍